dbms/src/Common/ThreadPool.h

```diff
@@ -0,0 +1,203 @@
+#include <functional>
+#include <queue>
+#include <list>
+#include <optional>
+#include <ext/singleton.h>
+
+#include <Common/ThreadStatus.h>
+
+
+/** Very simple thread pool similar to boost::threadpool.
+  * Advantages:
+  * - catches exceptions and rethrows on wait.
+  *
+  * This thread pool can be used as a task queue.
+  * For example, you can create a thread pool with 10 threads (and queue of size 10) and schedule 1000 tasks
+  * - in this case you will be blocked to keep 10 tasks in fly.
+  *
+  * Thread: std::thread or something with identical interface.
+  */
+template <typename Thread>
+class ThreadPoolImpl
+{
+public:
+    using Job = std::function<void()>;
+
+    /// Size is constant. Up to num_threads are created on demand and then run until shutdown.
+    explicit ThreadPoolImpl(size_t max_threads);
+
+    /// queue_size - maximum number of running plus scheduled jobs. It can be greater than max_threads. Zero means unlimited.
+    ThreadPoolImpl(size_t max_threads, size_t max_free_threads, size_t queue_size);
+
+    /// Add new job. Locks until number of scheduled jobs is less than maximum or exception in one of threads was thrown.
+    /// If an exception in some thread was thrown, method silently returns, and exception will be rethrown only on call to 'wait' function.
+    /// Priority: greater is higher.
+    void schedule(Job job, int priority = 0);
+
+    /// Wait for specified amount of time and schedule a job or return false.
+    bool trySchedule(Job job, int priority = 0, uint64_t wait_microseconds = 0);
+
+    /// Wait for specified amount of time and schedule a job or throw an exception.
+    void scheduleOrThrow(Job job, int priority = 0, uint64_t wait_microseconds = 0);
+
+    /// Wait for all currently active jobs to be done.
+    /// You may call schedule and wait many times in arbitary order.
+    /// If any thread was throw an exception, first exception will be rethrown from this method,
+    ///  and exception will be cleared.
+    void wait();
+
+    /// Waits for all threads. Doesn't rethrow exceptions (use 'wait' method to rethrow exceptions).
+    /// You should not destroy object while calling schedule or wait methods from another threads.
+    ~ThreadPoolImpl();
+
+    /// Returns number of running and scheduled jobs.
+    size_t active() const;
+
+private:
+    mutable std::mutex mutex;
+    std::condition_variable job_finished;
+    std::condition_variable new_job_or_shutdown;
+
+    const size_t max_threads;
+    const size_t max_free_threads;
+    const size_t queue_size;
+
+    size_t scheduled_jobs = 0;
+    bool shutdown = false;
+
+    struct JobWithPriority
+    {
+        Job job;
+        int priority;
+
+        JobWithPriority(Job job, int priority)
+            : job(job), priority(priority) {}
+
+        bool operator< (const JobWithPriority & rhs) const
+        {
+            return priority < rhs.priority;
+        }
+    };
+
+    std::priority_queue<JobWithPriority> jobs;
```

Comment:
Why use a shared-work strategy ThreadPool?

Reply:
Answered in Telegram.